### PR TITLE
Make consumer create target directories

### LINKED ISSUE #651: Consumer does not create directories for PAPERLESS_FILENAME_FORMAT
Coming from #593 ... the model.py does create the directories if needed, the consumer doesn't. Leads to no problems when the server is in charge of changing the name - but to a crash when the consumer consumes a new file.

### PR DESCRIPTION
**Pull request by BastianPoe**
Fixes #651

### CODE REVIEW

File: src/documents/tests/test_consumer.py [hunk 1]
```diff
@@ -1,14 +1,83 @@
...
+
+            # Check if consumed file has been stored correctly
+            self.assertEqual(os.path.isfile(os.path.join(
+                settings.MEDIA_ROOT, "documents", "originals", "none",
+                "letter-0000001.pdf.gpg")), True)
+
+    @override_settings(PAPERLESS_FILENAME_FORMAT="{correspondent}/dummy")
+    def test_duplicate_file_consumption(self):
+        with TemporaryDirectory() as tmpdir:
+            myConsumer = Consumer(consume=tmpdir)
```
Comment (MasterofJOKers) [inline src/documents/tests/test_consumer.py:58]: Shouldn't we override `setting.SCRATCH_DIR` too or pass it to the consumer with `scratch=` here? Otherwise, we would create the default `/tmp/paperless` and might leave some files behind if we die.

File: src/documents/tests/test_consumer.py [hunk 2]
```diff
@@ -1,14 +1,83 @@
...
+
+            # Create documents and thumbnails
+            os.makedirs(os.path.join(
+                settings.MEDIA_ROOT, "documents", "originals"), exist_ok=True)
+            os.makedirs(os.path.join(
+                settings.MEDIA_ROOT, "documents", "thumbnails"), exist_ok=True)
```
Comment (MasterofJOKers) [inline src/documents/tests/test_consumer.py:64]: That's already done in `setUp()`.

File: src/documents/models.py [hunk 3]
```diff
@@ -286,6 +286,10 @@ def find_renamed_document(self, subdirectory=""):
         # Go up in the directory hierarchy and try to delete all directories
```
Comment (MasterofJOKers) [inline src/documents/models.py:286]: We actually don't try to delete anything, but just search for a file.

File: src/documents/models.py [hunk 4]
```diff
@@ -286,6 +286,10 @@ def find_renamed_document(self, subdirectory=""):
...
         root = os.path.normpath(Document.filename_to_path(subdirectory))
 
+        # Check if root really exists and return otherwise
+        if not os.path.isdir(root):
+            return None
```
Comment (MasterofJOKers) [inline src/documents/models.py:291]: I'm a little confused here. Why is that necessary? Why didn't we pick up on this earlier? Could you explain a little - in the commit-message, if possible.

File: src/documents/tests/test_consumer.py [hunk 5]
```diff
@@ -1,14 +1,83 @@
 import re
 
-from django.test import TestCase
+from django.test import TestCase, override_settings
+from uuid import uuid4
+import os
+import shutil
 from unittest import mock
 from tempfile import TemporaryDirectory
+from django.conf import settings
```
[the remark this answers is not in the document]
Comment (BastianPoe) [reply src/documents/tests/test_consumer.py]: Done

File: src/documents/tests/test_consumer.py [hunk 6]
```diff
@@ -1,14 +1,83 @@
...
 
 class TestConsumer(TestCase):
+    SAMPLE_FILES = os.path.join(os.path.dirname(__file__), "samples")
+    deletion_list = []
+
+    def add_to_deletion_list(self, dirname):
+        self.deletion_list.append(dirname)
+
+    def setUp(self):
+        storage = "/tmp/paperless-tests-{}".format(str(uuid4())[:8])
```
[the remark this answers is not in the document]
Comment (BastianPoe) [reply src/documents/tests/test_consumer.py]: That is a relict that I copied from other tests (and can still be found in some other files). I removed this and went for full TemporaryDirectory.

File: src/documents/tests/test_consumer.py (hunk 2 above)
Comment (BastianPoe) [reply src/documents/tests/test_consumer.py]: Removed.

File: src/documents/models.py (hunk 4 above)
Comment (BastianPoe) [reply src/documents/models.py:292]: Turns out, that override_settings is actually keeping the override for
    following test cases (even if you disable them). This leads to an
    overriden MEDIA_ROOT that no longer exists. test_file_deletion does not
    do any file operations, but find_renamed_document was assuming the
    MEDIA_ROOT is actually exist.

This is actually pretty ugly and I am open for better suggestions.

File: src/documents/tests/test_consumer.py (hunk 1 above)
Comment (BastianPoe) [reply src/documents/tests/test_consumer.py]: Yep, good idea. Unfortunately, this required a number of follow-up changes in other files (see commits) but in general I think it is good cleaning this up properly.

File: src/documents/models.py (hunk 3 above)
Comment (BastianPoe) [reply src/documents/models.py]: fixed.

**Comment by tido-** [comment]
@MasterofJOKers   will you kindly spend 5 minutes?
grafik

**Comment by BastianPoe** [comment]
@MasterofJOKers @pitkley @danielquinn, could anyone of you please review this PR? It is open for several month now and I would like to close this issue. This is time-critical, as this addresses a bug currently found on master.r

**Review by pitkley — APPROVED** [review]

**Review by MasterofJOKers — APPROVED** [review]
